aip/0162.md

```diff
@@ -0,0 +1,312 @@
+dealing with the revision history (such as rollback) where failing to require
+it would not make sense (or lead to dangerous mistakes).
+
+### Getting a revision
+
+APIs implementing resource revisions **should** accept a resource name with a
+revision ID in the standard `Get` method ([AIP-131][]):
+
+```proto
+message GetBookRequest {
+  // The name of the book.
+  //   Example: publishers/123/books/les-miserables
+  //
+  // In order to retrieve a previous revision of the book, also provide
+  // the revision ID.
+  //   Example: publishers/123/books/les-miserables@c7cfa2a8
+  string name = 1;
+}
+```
+
+If the user passes a revision ID that does not exist, the API **must** fail
+with a `NOT_FOUND` error.
+
+APIs **must** return a `name` value corresponding to what the user sent. If the
+user sent a name with no revision ID, the returned `name` string **must not**
+include the revision ID either. Similarly, if the user sent a name with a
+revision ID, the returned `name` string **must** explicitly include it.
+
+APIs **should** allow users to specify a `@` with no revision ID to indicate
+that the user wants the latest revision, but for the revision ID to be
+explicitly included in the `name` field in the response:
+
+    publishers/123/books/les-miserables@
+
+### Listing revisions
+
+APIs implementing resource revisions **should** provide a custom method for
+listing the revision history for a resource, with a structure similar to
+standard `List` methods ([AIP-132][]):
+
+```proto
+rpc ListBookRevisions(ListBookRevisionsRequest)
+    returns (ListBookRevisionsResponse) {
+  option (google.api.http) = {
+    get: "/v1/{name=publishers/*/books/*}:listRevisions"
+  }
+}
+```
+
+```proto
+message ListBookRevisionsRequest {
+  // The name of the book to list revisions for.
+  string name = 1;
+
+  // The maximum number of revisions to return per page.
+  int32 page_size = 2;
+
+  // The page token, received from a previous ListBookRevisions call.
+  // Provide this to retrieve the subsequent page.
+  string page_token = 3;
+}
+
+message ListBookRevisionsResponse {
+  // The revisions of the book.
+  repeated Book books = 1;
+
+  // A token that can be sent as `page_token` to retrieve the next page.
+  // If this field is omitted, there are no subsequent pages.
+  string next_page_token = 2;
+}
+```
+
+While revision listing methods are mostly similar to standard `List` methods
+([AIP-132][]), the following important differences apply:
+
+- The first field in the request message **must** be called `name` rather than
+  `parent` (this is listing revisions for a specific book, not a collection of
+  books).
+- The URI **must** end with `:listRevisions`.
+- Revisions **must** be ordered in reverse chronological order.
+- The returned resources **must** include an explicit revision ID in the
+  resource's `name` field.
+  - If providing the full resource is expensive or infeasible, the revision
```

Comment:
The API **must** document that it will do this.

Reply:
Done.